mlx/dist/mpi/mpi.cpp

```diff
@@ -0,0 +1,114 @@
+// Copyright © 2024 Apple Inc.
+
+#include <mpi.h>
+
+#include "mlx/dist/dist.h"
+#include "mlx/scheduler.h"
+
+namespace mlx::core::dist {
+
+namespace {
+
+struct MPIGroup : public Group {
+  MPIGroup() : is_global(true) {
+    comm = MPI_COMM_WORLD;
+    MPI_Comm_size(comm, &size_);
+    MPI_Comm_rank(comm, &rank_);
+  }
+
+  ~MPIGroup() {
+    if (is_global) {
+      MPI_Finalize();
+    }
+  }
+
+  virtual int rank() override {
+    return rank_;
+  }
+
+  virtual int size() override {
+    return size_;
+  }
+
+  virtual std::shared_ptr<Group> split(int n) override {
+    throw std::runtime_error("MPIGroup split not yet implemented");
+  }
+
+  bool is_global;
+  MPI_Comm comm;
+  int rank_;
+  int size_;
+};
+
+MPI_Datatype mpi_datatype(const array& arr) {
+  switch (arr.dtype()) {
+    case bool_:
+      return MPI_C_BOOL;
+    case int8:
+      return MPI_INT8_T;
+    case uint8:
+      return MPI_UINT8_T;
+    case int16:
+      return MPI_INT16_T;
+    case uint16:
+      return MPI_UINT16_T;
+    case int32:
+      return MPI_INT32_T;
+    case uint32:
+      return MPI_UINT32_T;
+    case int64:
+      return MPI_INT64_T;
+    case uint64:
+      return MPI_UINT64_T;
+    case float32:
+      return MPI_FLOAT;
+    case complex64:
+      return MPI_C_COMPLEX;
+    case float16:
+    case bfloat16:
+      throw std::runtime_error("MPI doesn't support 16-bit floats");
+  }
+}
+
+} // namespace
+
+bool is_available() {
+  return true;
+}
+
+Stream communication_stream() {
+  static std::shared_ptr<Stream> comm_stream = nullptr;
+  if (comm_stream == nullptr) {
+    comm_stream = std::make_shared<Stream>(
+        scheduler::scheduler().new_stream(Device::cpu));
+  }
```

Comment:
nit: maybe a unique_ptr is better here

Reply:
Nit: Also m use the free function `new_stream(Device::cpu)` rather than call the scheduler directly

Reply:
The only reason I did the heap allocation was to avoid potential pitfalls with static initializations. I can do the above instead and reconsider if we encounter issues. I am also wondering what would happen if I used an optional initialized empty 🤔

Reply:
I think it should be ok.. either way it won't be initialized until this function is called the first time which will determine destruction order.